web/js/codeworld_shared.js

```diff
@@ -122,95 +122,161 @@ var hintBlacklist = [
-    if (data && data.token && data.token.string) {
-        var token_name = data.token.string;
-        if (hintBlacklist.indexOf(token_name) == -1 && codeWorldBuiltinsDocs[token_name]){
-            return codeWorldBuiltinsDocs[token_name];
-        }
+function getWordEnd(word, line) {
+    var wordStart = getWordStart(word, line);
+    if (wordStart != -1) {
+        return wordStart + word.length
     }
-    return;
+    return -1;
 }
 
-// Hints and hover tooltips
-function registerStandardHints(successFunc)
-{
-    function createHint(line, wordStart, wordEnd, cname) {
-        var word = line.slice(wordStart, wordEnd);
-        if (!cname) cname = 'hint-word';
+function parseSymbolsFromCurrentCode() {
+    var lines = window.codeworldEditor.getValue().split('\n'),
+        word = null,
+        parseResults = {};
 
-        function renderer(elem, data, cur) {
-            if (wordStart > 0) {
-                elem.appendChild(document.createTextNode(line.slice(0, wordStart)));
+    lines.forEach(function(line) {
+        // f(x, y) =
+        if (/^\w+\(.*/.test(line)) {
+            word = line.split("(")[0];
+            parseResults[word] = {
+                symbolStart: getWordStart(word, line),
+                symbolEnd: getWordEnd(word, line)
             }
-            var wordElem = document.createElement("span");
-            wordElem.className = cname;
-            wordElem.appendChild(document.createTextNode(word));
-            elem.appendChild(wordElem);
-            if (wordEnd < line.length) {
-                var leftover = line.slice(wordEnd).replace(/\s+/g, ' ');
-                if (wordEnd + leftover.length > 60 && leftover.length > 3) {
-                  leftover = leftover.slice(0, 57 - wordEnd) + '...';
-                }
-                elem.appendChild(document.createTextNode(leftover));
-                elem.title = line;
+        }
+        // foo =
+        else if (/^\S+\s*=/.test(line)) {
+            word = line.split("=")[0].trim();
+            parseResults[word] = {
+                symbolStart: getWordStart(word, line),
+                symbolEnd: getWordEnd(word, line)
+            };
+        }
+        // data Foo
+        else if (/^data\s.+/.test(line)) {
+            word = line.split(" ")[1];
+            parseResults[word] = {
+                declaration: line.slice(0, getWordStart(word, line)),
+                symbolStart: getWordStart(word, line),
+                symbolEnd: getWordEnd(word, line)
             }
         }
-        return {
-            text: word,
-            render: renderer,
-            source: line
-        };
-    }
-
-    function createHover(line, wordStart, wordEnd, doc, hint){
-        if (!hint){
-            hint = createHint(line, wordStart, wordEnd);
+        // type Foo = Bar
+        else if (/^type\s.+/.test(line)) {
+            var splitted = line.split("=");
+            word = splitted[0].trim().split(" ").slice(-1)[0];
```

Comment:
Same as above: use capturing groups to extract the name.  This won't work for `type Foo=Bar` for example.  I won't keep repeating this; hopefully you get the idea.

Reply:
Done.